scipy/stats/_distn_infrastructure.py

```diff
@@ -3229,36 +3225,60 @@ def fun(x):
         else:
             invfac = 1.0
 
-        tot = 0.0
-        low, upp = self._ppf(0.001, *args), self._ppf(0.999, *args)
-        low = max(min(-suppnmin, low), lb)
-        upp = min(max(suppnmin, upp), ub)
-        supp = np.arange(low, upp+1, self.inc)  # check limits
-        tot = np.sum(fun(supp))
-        diff = 1e100
-        pos = upp + self.inc
-        count = 0
+        # short-circuit if the support size is small enough
+        if (ub - lb) <= chunksize:
+            supp = np.arange(lb, ub+1, self.inc)
+            vals = fun(supp)
+            return np.sum(vals) / invfac
+
+        # otherwise, iterate starting from median
+        x0 = self.ppf(0.5, *args)
+        if x0 < lb:
+            x0 = lb
+        if x0 > ub:
+            x0 = ub
+
+        count, tot = 0, 0.
+        # iterate over [x0, ub] inclusive
+        for x in _iter_chunked(x0, ub+1, chunksize=chunksize, inc=self.inc):
+            count += x.size
+            delta = np.sum(fun(x))
+            tot += delta
+            if abs(delta) < tolerance * x.size:
+                break
+            if count > maxcount:
+                warnings.warn('expect(): sum did not converge', RuntimeWarning)
+                return tot / invfac
+        
+        # iterate over [lb, x0)
+        for x in _iter_chunked(x0-1, lb-1, chunksize=chunksize, inc=-self.inc):
+            count += x.size
+            delta = np.sum(fun(x))
+            tot += delta
+            if abs(delta) < tolerance * x.size:
+                break
+            if count > maxcount:
+                warnings.warn('expect(): sum did not converge', RuntimeWarning)
+                break
 
-        # handle cases with infinite support
+        return tot/invfac
 
-        while (pos <= ub) and (diff > self.moment_tol) and count <= maxcount:
-            diff = fun(pos)
-            tot += diff
-            pos += self.inc
-            count += 1
 
-        if self.a < 0:  # handle case when self.a = -inf
-            diff = 1e100
-            pos = low - self.inc
-            while ((pos >= lb) and (diff > self.moment_tol) and
-                   count <= maxcount):
-                diff = fun(pos)
-                tot += diff
-                pos -= self.inc
-                count += 1
-        if count > maxcount:
-            warnings.warn('expect(): sum did not converge', RuntimeWarning)
-        return tot/invfac
+def _iter_chunked(x0, x1, chunksize=4, inc=1):
+    """Iterate from x0 to x1 *inclusive* in chunks of chunksize and steps inc.
```

Comment:
I was just looking at my trial run for this function again.
`*inclusive*` is wrong, isn't it?
it does not include the end point x1, AFAICS, you adjust x1 in the call to this function

(I might reuse a copy of this iterator, so I was paying more attention to it as standalone function.)